crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,411 @@
+                bcs::to_bytes(contents)?,
+            ),
+        ];
+        self.multi_set(CHECKPOINTS_TABLE, [(key, cells)]).await
+    }
+}
+
+#[async_trait]
+impl KeyValueStoreReader for BigTableClient {
+    async fn get_objects(&mut self, object_keys: &[ObjectKey]) -> Result<Vec<Object>> {
+        let keys: Result<_, _> = object_keys.iter().map(Self::raw_object_key).collect();
+        let mut objects = vec![];
+        for row in self.multi_get(OBJECTS_TABLE, keys?).await? {
+            for (_, value) in row {
+                objects.push(bcs::from_bytes(&value)?);
+            }
+        }
+        Ok(objects)
+    }
+
+    async fn get_transactions(
+        &mut self,
+        transactions: &[TransactionDigest],
+    ) -> Result<Vec<TransactionData>> {
+        let keys = transactions.iter().map(|tx| tx.inner().to_vec()).collect();
+        let mut result = vec![];
+        for row in self.multi_get(TRANSACTIONS_TABLE, keys).await? {
+            let mut transaction = None;
+            let mut effects = None;
+            let mut events = None;
+            let mut timestamp = 0;
+            let mut checkpoint_number = 0;
+
+            for (column, value) in row {
+                match std::str::from_utf8(&column)? {
+                    TRANSACTION_COLUMN_QUALIFIER => transaction = Some(bcs::from_bytes(&value)?),
+                    EFFECTS_COLUMN_QUALIFIER => effects = Some(bcs::from_bytes(&value)?),
+                    EVENTS_COLUMN_QUALIFIER => events = Some(bcs::from_bytes(&value)?),
+                    TIMESTAMP_COLUMN_QUALIFIER => timestamp = bcs::from_bytes(&value)?,
+                    CHECKPOINT_NUMBER_COLUMN_QUALIFIER => {
+                        checkpoint_number = bcs::from_bytes(&value)?
+                    }
+                    _ => error!("unexpected column {:?} in transactions table", column),
+                }
+            }
+            result.push(TransactionData {
+                transaction: transaction.ok_or_else(|| anyhow!("transaction field is missing"))?,
+                effects: effects.ok_or_else(|| anyhow!("effects field is missing"))?,
+                events: events.ok_or_else(|| anyhow!("events field is missing"))?,
+                timestamp,
+                checkpoint_number,
+            })
+        }
+        Ok(result)
+    }
+
+    async fn get_checkpoint(
+        &mut self,
+        sequence_number: SequenceNumber,
+    ) -> Result<Option<Checkpoint>> {
+        let key = sequence_number.value().to_be_bytes().to_vec();
+        if let Some(cells) = self.multi_get(CHECKPOINTS_TABLE, vec![key]).await?.pop() {
+            let (mut summary, mut contents) = (None, None);
+            for (column, value) in cells {
+                match std::str::from_utf8(&column)? {
+                    CHECKPOINT_SUMMARY_COLUMN_QUALIFIER => summary = bcs::from_bytes(&value)?,
+                    CHECKPOINT_CONTENTS_COLUMN_QUALIFIER => contents = bcs::from_bytes(&value)?,
```

Comment:
I sort of would like to have the signature separated from the summary itself in their own column, mainly because the summary is explicitly fixed while the signature could technically vary depending on when signatures were witnessed/aggregated.

Reply:
done